caas/kubernetes/provider/application/constraints.go

```diff
@@ -266,35 +315,139 @@ func processPodAffinity(pod *core.PodSpec, affinityLabels map[string]string) err
 	return nil
 }
 
-func configureConstraint(pod *core.PodSpec, resourceName core.ResourceName, value string) (err error) {
+// divideAndSpreadContainerResource evenly distributes the total resource across a given number of containers.
+// It returns a slice of len(containers), where each element represents the distributed value
+// allocated to that container. Any remainder is distributed one-by-one starting from the lowest index.
+// e.g., [4,4,3] for totalResource = 11, numContainers = 3
+func divideAndSpreadContainerResource(totalResource uint64, numContainers int) []uint64 {
+	if numContainers <= 0 {
+		return nil
+	}
+
+	result := make([]uint64, numContainers)
+	quotient := totalResource / uint64(numContainers)
+	remainder := totalResource % uint64(numContainers)
+
+	for i := 0; i < numContainers; i++ {
+		if uint64(i) < remainder {
+			result[i] = quotient + 1
+		} else {
+			result[i] = quotient
+		}
+	}
+	return result
+}
+
+// configureWorkloadConstraint sets the resource limits for all containers,
+// and the resource request for the first workload container only.
+func configureWorkloadConstraint(pod *core.PodSpec, resourceName core.ResourceName, value uint64) (err error) {
 	if len(pod.Containers) == 0 {
 		return nil
 	}
-	pod.Containers[0].Resources.Requests, err = MergeConstraint(resourceName, value, pod.Containers[0].Resources.Requests)
+	var unitSuffix string
+
+	switch resourceName {
+	case core.ResourceMemory:
+		unitSuffix = "Mi"
+	case core.ResourceCPU:
+		unitSuffix = "m"
+	default:
+		return errors.NotSupportedf("converting resource value for %q", resourceName)
+	}
+
+	resourceValStr := fmt.Sprintf("%d%s", value, unitSuffix)
+	resourceQty, err := resource.ParseQuantity(resourceValStr)
 	if err != nil {
-		return errors.Annotatef(err, "merging request constraint %s=%s", resourceName, value)
+		return errors.Annotatef(err, "parsing constraint value %q for %q", value, resourceName)
+	}
+	pod.Containers[0].Resources.Requests, err = MergeConstraint(resourceName, resourceQty, pod.Containers[0].Resources.Requests)
+	if err != nil {
+		return errors.Annotatef(err, "merging request constraint %s=%s", resourceName, resourceValStr)
 	}
 	for i := range pod.Containers {
-		pod.Containers[i].Resources.Limits, err = MergeConstraint(resourceName, value, pod.Containers[i].Resources.Limits)
+		pod.Containers[i].Resources.Limits, err = MergeConstraint(resourceName, resourceQty, pod.Containers[i].Resources.Limits)
+		if err != nil {
+			return errors.Annotatef(err, "merging limit constraint %s=%s", resourceName, resourceValStr)
```

Comment:
Might be good to include the container name in the error.

Reply:
refactored w thanks